src/utl/src/Progress.cpp

```diff
@@ -0,0 +1,332 @@
+/////////////////////////////////////////////////////////////////////////////
+//
+// Copyright (c) 2025, The Regents of the University of California
+// All rights reserved.
+//
+// BSD 3-Clause License
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#include "utl/Progress.h"
+
+#include "utl/Logger.h"
+
+namespace utl {
+
+static Progress::ProgressHalt* signal_halt = nullptr;
+static void controlC_handler(int sig)
+{
+  if (signal_halt == nullptr) {
+    return;
+  }
+
+  const auto now = std::time(nullptr);
+  signal_halt->interrupts.push(now);
+
+  // Remove old interrupts
+  while (std::difftime(now, signal_halt->interrupts.front())
+         >= Progress::ProgressHalt::max_prune_delay) {
+    signal_halt->interrupts.pop();
+  }
+
+  if (signal_halt->interrupts.size()
+      >= Progress::ProgressHalt::max_interrupts) {
+    // halt the application
+    signal(sig, SIG_DFL);
+    raise(sig);
+  } else if (signal_halt->interrupts.size() + 1
+             >= Progress::ProgressHalt::max_interrupts) {
+    // let user know one more will terminate openroad
```

Comment:
If there is no process to interrupt then it would be good to have a message on the first ctrl-c.  Currently it just disappears with no response.

Reply:
should be easy